diff --git a/.gitignore b/.gitignore
index b436353..5ccecbc 100644
--- a/.gitignore
+++ b/.gitignore
@@ -57,3 +57,5 @@ deps/openssl/openssl.xml
 /SHASUMS.txt*
 
 /tools/wrk/wrk
+
+*.xcodeproj
diff --git a/Makefile b/Makefile
index b0d3375..5b7b078 100644
--- a/Makefile
+++ b/Makefile
@@ -34,20 +34,32 @@ endif
 
 ifeq ($(USE_NINJA),1)
 node: config.gypi
-	$(NINJA) -C out/Release/
+	$(NINJA) -C out/Release/ node
 	ln -fs out/Release/node node
 
 node_g: config.gypi
 	$(NINJA) -C out/Debug/
 	ln -fs out/Debug/node $@
+
+staticlib: config.gypi
+	$(NINJA) -C out/Release/ node_base
+
+staticlib_g: config.gypi
+	$(NINJA) -C out/Debug/ node_base
 else
 node: config.gypi out/Makefile
-	$(MAKE) -C out BUILDTYPE=Release V=$(V)
+	$(MAKE) -C out BUILDTYPE=Release V=$(V) node
 	ln -fs out/Release/node node
 
 node_g: config.gypi out/Makefile
 	$(MAKE) -C out BUILDTYPE=Debug V=$(V)
 	ln -fs out/Debug/node $@
+
+staticlib: config.gypi
+	$(MAKE) -C out BUILDTYPE=Release V=$(V) node_base
+
+staticlib_g: config.gypi
+	$(MAKE) -C out BUILDTYPE=Debug V=$(V) node_base
 endif
 
 out/Makefile: common.gypi deps/uv/uv.gyp deps/http_parser/http_parser.gyp deps/zlib/zlib.gyp deps/v8/build/toolchain.gypi deps/v8/build/features.gypi deps/v8/tools/gyp/v8.gyp node.gyp config.gypi
diff --git a/deps/openssl/openssl/crypto/bn/bn_div.c b/deps/openssl/openssl/crypto/bn/bn_div.c
index 7b24031..5f24a34 100644
--- a/deps/openssl/openssl/crypto/bn/bn_div.c
+++ b/deps/openssl/openssl/crypto/bn/bn_div.c
@@ -143,7 +143,7 @@ int BN_div(BIGNUM *dv, BIGNUM *rem, const BIGNUM *m, const BIGNUM *d,
     */
 #undef bn_div_words
 #  define bn_div_words(n0,n1,d0)		\
-	({  asm volatile (			\
+	({  __asm__ volatile (			\
 		"divl	%4"			\
 		: "=a"(q), "=d"(rem)		\
 		: "a"(n1), "d"(n0), "g"(d0)	\
@@ -158,7 +158,7 @@ int BN_div(BIGNUM *dv, BIGNUM *rem, const BIGNUM *m, const BIGNUM *d,
     */
 #  undef bn_div_words
 #  define bn_div_words(n0,n1,d0)		\
-	({  asm volatile (			\
+	({  __asm__ volatile (			\
 		"divq	%4"			\
 		: "=a"(q), "=d"(rem)		\
 		: "a"(n1), "d"(n0), "g"(d0)	\
diff --git a/deps/openssl/openssl/crypto/md32_common.h b/deps/openssl/openssl/crypto/md32_common.h
index bb73819..201097b 100644
--- a/deps/openssl/openssl/crypto/md32_common.h
+++ b/deps/openssl/openssl/crypto/md32_common.h
@@ -162,7 +162,7 @@
    */
 #  if defined(__i386) || defined(__i386__) || defined(__x86_64) || defined(__x86_64__)
 #   define ROTATE(a,n)	({ register unsigned int ret;	\
-				asm (			\
+				__asm__ (			\
 				"roll %1,%0"		\
 				: "=r"(ret)		\
 				: "I"(n), "0"((unsigned int)(a))	\
@@ -172,7 +172,7 @@
 #  elif defined(_ARCH_PPC) || defined(_ARCH_PPC64) || \
 	defined(__powerpc) || defined(__ppc__) || defined(__powerpc64__)
 #   define ROTATE(a,n)	({ register unsigned int ret;	\
-				asm (			\
+				__asm__ (			\
 				"rlwinm %0,%1,%2,0,31"	\
 				: "=r"(ret)		\
 				: "r"(a), "I"(n));	\
@@ -180,7 +180,7 @@
 			})
 #  elif defined(__s390x__)
 #   define ROTATE(a,n) ({ register unsigned int ret;	\
-				asm ("rll %0,%1,%2"	\
+				__asm__ ("rll %0,%1,%2"	\
 				: "=r"(ret)		\
 				: "r"(a), "I"(n));	\
 			  ret;				\
@@ -207,10 +207,10 @@
      * unaligned data without raising an exception.
      */
 #   define HOST_c2l(c,l)	({ unsigned int r=*((const unsigned int *)(c));	\
-				   asm ("bswapl %0":"=r"(r):"0"(r));	\
+				   __asm__ ("bswapl %0":"=r"(r):"0"(r));	\
 				   (c)+=4; (l)=r;			})
 #   define HOST_l2c(l,c)	({ unsigned int r=(l);			\
-				   asm ("bswapl %0":"=r"(r):"0"(r));	\
+				   __asm__ ("bswapl %0":"=r"(r):"0"(r));	\
 				   *((unsigned int *)(c))=r; (c)+=4; r;	})
 #   endif
 #  endif
@@ -241,10 +241,10 @@
 #ifndef PEDANTIC
 # if defined(__GNUC__) && __GNUC__>=2 && !defined(OPENSSL_NO_ASM) && !defined(OPENSSL_NO_INLINE_ASM)
 #  if defined(__s390x__)
-#   define HOST_c2l(c,l)	({ asm ("lrv	%0,%1"			\
+#   define HOST_c2l(c,l)	({ __asm__ ("lrv	%0,%1"			\
 				   :"=d"(l) :"m"(*(const unsigned int *)(c)));\
 				   (c)+=4; (l);				})
-#   define HOST_l2c(l,c)	({ asm ("strv	%1,%0"			\
+#   define HOST_l2c(l,c)	({ __asm__ ("strv	%1,%0"			\
 				   :"=m"(*(unsigned int *)(c)) :"d"(l));\
 				   (c)+=4; (l);				})
 #  endif
diff --git a/deps/openssl/openssl/crypto/modes/modes_lcl.h b/deps/openssl/openssl/crypto/modes/modes_lcl.h
index b6dc3c3..ab1110e 100644
--- a/deps/openssl/openssl/crypto/modes/modes_lcl.h
+++ b/deps/openssl/openssl/crypto/modes/modes_lcl.h
@@ -40,26 +40,26 @@ typedef unsigned char u8;
 #if defined(__GNUC__) && __GNUC__>=2
 # if defined(__x86_64) || defined(__x86_64__)
 #  define BSWAP8(x) ({	u64 ret=(x);			\
-			asm ("bswapq %0"		\
+			__asm__ ("bswapq %0"		\
 			: "+r"(ret));	ret;		})
 #  define BSWAP4(x) ({	u32 ret=(x);			\
-			asm ("bswapl %0"		\
+			__asm__ ("bswapl %0"		\
 			: "+r"(ret));	ret;		})
 # elif (defined(__i386) || defined(__i386__)) && !defined(I386_ONLY)
 #  define BSWAP8(x) ({	u32 lo=(u64)(x)>>32,hi=(x);	\
-			asm ("bswapl %0; bswapl %1"	\
+			__asm__ ("bswapl %0; bswapl %1"	\
 			: "+r"(hi),"+r"(lo));		\
 			(u64)hi<<32|lo;			})
 #  define BSWAP4(x) ({	u32 ret=(x);			\
-			asm ("bswapl %0"		\
+			__asm__ ("bswapl %0"		\
 			: "+r"(ret));	ret;		})
 # elif (defined(__arm__) || defined(__arm)) && !defined(STRICT_ALIGNMENT)
 #  define BSWAP8(x) ({	u32 lo=(u64)(x)>>32,hi=(x);	\
-			asm ("rev %0,%0; rev %1,%1"	\
+			__asm__ ("rev %0,%0; rev %1,%1"	\
 			: "+r"(hi),"+r"(lo));		\
 			(u64)hi<<32|lo;			})
 #  define BSWAP4(x) ({	u32 ret;			\
-			asm ("rev %0,%1"		\
+			__asm__ ("rev %0,%1"		\
 			: "=r"(ret) : "r"((u32)(x)));	\
 			ret;				})
 # endif
diff --git a/node.gyp b/node.gyp
index d6aefae..a1a30d1 100644
--- a/node.gyp
+++ b/node.gyp
@@ -73,17 +73,14 @@
 
   'targets': [
     {
-      'target_name': 'node',
-      'type': 'executable',
-
+      'target_name': 'node_base',
+      'type': 'static_library',
       'dependencies': [
         'node_js2c#host',
       ],
 
       'include_dirs': [
         'src',
-        'tools/msvs/genfiles',
-        'deps/uv/src/ares',
         '<(SHARED_INTERMEDIATE_DIR)' # for node_natives.h
       ],
 
@@ -98,7 +95,6 @@
         'src/node_file.cc',
         'src/node_http_parser.cc',
         'src/node_javascript.cc',
-        'src/node_main.cc',
         'src/node_os.cc',
         'src/node_script.cc',
         'src/node_stat_watcher.cc',
@@ -168,10 +164,14 @@
             'src/tls_wrap.cc',
             'src/tls_wrap.h'
           ],
+
           'conditions': [
             [ 'node_shared_openssl=="false"', {
-              'dependencies': [ './deps/openssl/openssl.gyp:openssl' ],
-            }]]
+              'dependencies': [ 'deps/openssl/openssl.gyp:openssl' ],
+            }, {
+              'include_dirs': [ '<(shared_openssl_include_dir)' ],
+            }],
+          ]
         }, {
           'defines': [ 'HAVE_OPENSSL=0' ]
         }],
@@ -244,6 +244,7 @@
             'tools/msvs/genfiles/node_perfctr_provider.rc',
           ]
         } ],
+
         [ 'v8_postmortem_support=="true"', {
           'dependencies': [ 'deps/v8/tools/gyp/v8.gyp:postmortem-metadata' ],
         }],
@@ -253,22 +254,37 @@
             'deps/v8/include/v8-debug.h',
           ],
           'dependencies': [ 'deps/v8/tools/gyp/v8.gyp:v8' ],
+          'include_dirs': [ 'deps/v8/include' ],
+        }, {
+          'include_dirs': [ '<(shared_v8_include_dir)' ],
         }],
 
         [ 'node_shared_zlib=="false"', {
           'dependencies': [ 'deps/zlib/zlib.gyp:zlib' ],
+          'include_dirs': [ 'deps/zlib' ],
+        }, {
+          'include_dirs': [ '<(shared_zlib_include_dir)' ],
         }],
 
         [ 'node_shared_http_parser=="false"', {
           'dependencies': [ 'deps/http_parser/http_parser.gyp:http_parser' ],
+          'include_dirs': [ 'deps/http_parser' ],
+        }, {
+          'include_dirs': [ '<(shared_http_parser_include_dir)' ],
         }],
 
         [ 'node_shared_cares=="false"', {
           'dependencies': [ 'deps/cares/cares.gyp:cares' ],
+          'include_dirs': [ 'deps/cares/include' ],
+        }, {
+          'include_dirs': [ '<(shared_cares_include_dir)' ],
         }],
 
         [ 'node_shared_libuv=="false"', {
           'dependencies': [ 'deps/uv/uv.gyp:libuv' ],
+          'include_dirs': [ 'deps/uv/include' ],
+        }, {
+          'include_dirs': [ '<(shared_libuv_include_dir)' ],
         }],
 
         [ 'OS=="win"', {
@@ -295,6 +311,59 @@
             'PLATFORM="darwin"',
           ],
         }],
+      ],
+      'msvs_settings': {
+        'VCLinkerTool': {
+          'SubSystem': 1, # /subsystem:console
+        },
+      },
+    },
+
+    {
+      'target_name': 'node',
+      'type': 'executable',
+      'dependencies': [
+        'node_base',
+      ],
+      'include_dirs': [
+        'src',
+        'tools/msvs/genfiles',
+        'deps/uv/src/ares',
+      ],
+      'sources': [
+        'src/node_main.cc',
+      ],
+      'conditions': [
+        [ 'node_use_openssl=="true"', {
+          'conditions': [
+            [ 'node_shared_openssl=="false"', {
+              'dependencies': [ 'deps/openssl/openssl.gyp:openssl' ],
+            }],
+          ]
+        }],
+        [ 'v8_postmortem_support=="true"', {
+          'dependencies': [ 'deps/v8/tools/gyp/v8.gyp:postmortem-metadata' ],
+        }],
+        [ 'node_shared_v8=="false"', {
+          'dependencies': [ 'deps/v8/tools/gyp/v8.gyp:v8' ],
+        }],
+
+        [ 'node_shared_zlib=="false"', {
+          'dependencies': [ 'deps/zlib/zlib.gyp:zlib' ],
+        }],
+
+        [ 'node_shared_http_parser=="false"', {
+          'dependencies': [ 'deps/http_parser/http_parser.gyp:http_parser' ],
+        }],
+
+        [ 'node_shared_cares=="false"', {
+          'dependencies': [ 'deps/cares/cares.gyp:cares' ],
+        }],
+
+        [ 'node_shared_libuv=="false"', {
+          'dependencies': [ 'deps/uv/uv.gyp:libuv' ],
+        }],
+
         [ 'OS=="freebsd"', {
           'libraries': [
             '-lutil',
@@ -322,6 +391,7 @@
         },
       },
     },
+
     # generate ETW header and resource files
     {
       'target_name': 'node_etw',
diff --git a/src/node.cc b/src/node.cc
index 7b1d843..de5b4d0 100644
--- a/src/node.cc
+++ b/src/node.cc
@@ -3140,6 +3140,20 @@ static char **copy_argv(int argc, char **argv) {
   return argv_copy;
 }
 
+void SetupBindingCache() {
+  binding_cache.Reset(Isolate::GetCurrent(), Object::New());
+}
+
+void InitSetup(int argc, char *argv[]) {
+  // This needs to run *before* V8::Initialize()
+  // Use copy here as to not modify the original argv:
+  Init(argc, argv);
+
+  // Apparently we need to reassign this one
+  // TODO(spolu): figure out why?
+  node_isolate = Isolate::GetCurrent();
+}
+
 int Start(int argc, char *argv[]) {
   // Hack aroung with the argv pointer. Used for process.title = "blah".
   argv = uv_setup_args(argc, argv);
diff --git a/src/node.h b/src/node.h
index d7996fb..77712a8 100644
--- a/src/node.h
+++ b/src/node.h
@@ -120,10 +120,13 @@ NODE_EXTERN extern bool no_deprecation;
 
 NODE_EXTERN int Start(int argc, char *argv[]);
 
+void InitSetup(int argc, char *argv[]);
 char** Init(int argc, char *argv[]);
 v8::Handle<v8::Object> SetupProcessObject(int argc, char *argv[]);
+void SetupBindingCache();
 void Load(v8::Handle<v8::Object> process);
 void EmitExit(v8::Handle<v8::Object> process);
+void RunAtExit();
 
 /* Converts a unixtime to V8 Date */
 #define NODE_UNIXTIME_V8(t) v8::Date::New(1000*static_cast<double>(t))
